src/Features/Core/Portable/Snippets/AbstractConsoleSnippetProvider.cs

```diff
@@ -0,0 +1,125 @@
+            var symbol = compilation.GetTypeByMetadataName(typeof(Console).FullName);
+            if (symbol is null)
+            {
+                return false;
+            }
+
+            if (!ShouldDisplaySnippet(syntaxContext))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        protected override string GetSnippetDisplayName()
+        {
+            return FeaturesResources.Write_to_the_Console;
+        }
+
+        protected override async Task<ImmutableArray<TextChange>> GenerateSnippetTextChangesAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            using var _ = ArrayBuilder<TextChange>.GetInstance(out var arrayBuilder);
+            var snippetTextChange = await GenerateSnippetAsync(document, position, cancellationToken).ConfigureAwait(false);
+            arrayBuilder.AddRange(snippetTextChange);
+            return arrayBuilder.ToImmutableArray();
+        }
+
+        private async Task<TextChange> GenerateSnippetAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var symbol = await GetSymbolFromMetaDataNameAsync(document, cancellationToken).ConfigureAwait(false);
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var token = tree.FindTokenOnLeftOfPosition(position, cancellationToken);
+            var typeExpression = generator.TypeExpression(symbol.GetSymbolType());
+            var declaration = GetAsyncSupportingDeclaration(token);
+            var isAsync = generator.GetModifiers(declaration).IsAsync;
+            SyntaxNode? invocation;
+            invocation = isAsync
+                ? generator.ExpressionStatement(generator.AwaitExpression(generator.InvocationExpression(
+                    generator.MemberAccessExpression(generator.MemberAccessExpression(typeExpression, generator.IdentifierName(nameof(Console.Out))), generator.IdentifierName(nameof(Console.Out.WriteLineAsync))))))
+                : generator.ExpressionStatement(generator.InvocationExpression(generator.MemberAccessExpression(typeExpression, generator.IdentifierName(nameof(Console.WriteLine)))));
+            return new TextChange(TextSpan.FromBounds(position, position), invocation.NormalizeWhitespace().ToFullString());
+        }
+
+        protected override int GetTargetCaretPosition(ISyntaxFactsService syntaxFacts, SyntaxNode caretTarget)
+        {
+            var invocationExpression = caretTarget.DescendantNodes().Where(syntaxFacts.IsInvocationExpression).First();
+            var argumentListNode = syntaxFacts.GetArgumentListOfInvocationExpression(invocationExpression);
+            return argumentListNode!.GetLocation().SourceSpan.End - 1;
```

Comment:
how do yoiu know this is safe?

Reply:
1. you shouldn't need that `!`.  
2. add an ISyntaxFacts GetPartsOfArgumentList.  then use the close-paren that is returned and use the start of that (or the end of hte open-paren).